### PR TITLE
Rework attribute parsing

### PR DESCRIPTION
**Pull request by escritorio-gustavo**
## Goal

This PR aims to change the way we parse attributes in a few key ways:
- Introducing a couple of traits for the methods all attribute types share
- Change the `merge` method to remove the mutable reference
  - The reasoning for this is mostly that I think a `for_each` that mutates external data is somewhat confusing, so I changed it to use `fold` (think JS `Array.protoype.reduce`) instead.
- Bring all the attribute compatibility validation to the same file where parsing is done

## Checklist

- [x] I have followed the steps listed in the [Contributing guide](https://github.com/Aleph-Alpha/ts-rs/blob/main/CONTRIBUTING.md).
- [x] If necessary, I have added documentation related to the changes made.
- [ ] I have added or updated the tests related to the changes made.
  - This is a refactor, so there isn't really new anything I can test... We could make tests for whether or not our compiler errors are correct, but we still have to figure out how to do that

### CODE REVIEW

File: macros/src/attr/struct.rs [hunk 1]
```diff
@@ -28,67 +28,88 @@ pub struct SerdeStructAttr(StructAttr);
...
+        variant_fields: &Fields,
+    ) -> Self {
         Self {
             crate_rename: Some(enum_attr.crate_rename()),
             rename: variant_attr.rename.clone(),
-            rename_all: variant_attr.rename_all,
+            rename_all: variant_attr.rename_all.or(match variant_fields {
+                Fields::Named(_) => enum_attr.rename_all_fields,
+                Fields::Unnamed(_) | Fields::Unit => None,
+            }),
```
Comment (escritorio-gustavo) [inline macros/src/attr/struct.rs:53]: I managed to move the `rename_all_fields` match to a better location and get rid of that duplicated mehtod

**Comment by NyxCode** [comment]
I'll take a closer look once I get home, but I love what I've seen so far. Great work!

**Comment by NyxCode** [comment]
Great stuff, nothing here I don't like!  
As far as I'm concerned, you can merge this as-is - unless there's anything else you'd like to do.

**Comment by escritorio-gustavo** [comment]
> Great stuff, nothing here I don't like!

Right as you say that I push a compiler error lmao

**Comment by NyxCode** [comment]
What was the reasoning behind the last commit? Was there a bug I missed?

**Comment by escritorio-gustavo** [comment]
> What was the reasoning behind the last commit? Was there a bug I missed?

I thought it was weird passing the `ts` attribute into `parse_serde_attrs` for the initial value of `fold`, so i changed it to avoid doing that

**Comment by escritorio-gustavo** [comment]
> I think that was the last thing that I wanted to change, as long as you like the latest commit, I think this is ready to merge

Actually, I just realized that the `merge_with_serde` method will cause a pretty annoying difference, which is that to make a serde attr compatible with a ts attr `merge_with_serde` needs to be updated, instead of just the `impl_parse` block

**Comment by NyxCode** [comment]
Yeah, that's true. I too thought passing in the "previous" attrs was a tiny bit akward, but I think I preferred it over having the separate `merge_with_serde` fns.

**Comment by escritorio-gustavo** [comment]
In [ba04528](https://github.com/Aleph-Alpha/ts-rs/pull/290/commits/ba04528662ca13d300b674f5875e1340a672a689) removing `merge_with_serde` and using `merge` broke the merging of `docs`, so I just moved the `parse_docs` part to be after the serde attribute parsing

**Comment by escritorio-gustavo** [comment]
Alright, I think I'm happy with this now!

**Review by NyxCode — APPROVED** [review]
So am I!  
Great work, super happy you tackled this!

**Pull request merged**